Implement treegrid to display parent-child-relationship in project list

rbt-mm opened the pull request:
[Backend PR](https://github.com/rbt-mm/dependency-track/pull/8)

### Project list
Adds a treegrid to the project list table to display a hierarchical view, which shows the (already existing) multi-level parent-child-relationships between projects with lazy-loading in between levels.

The project list is hierarchical on default, but can be toggled to the old flat view.

Sorting and filtering by tags or classifiers works with the treegrid and the flat view, searching automatically switches to the flat view but preserves the state before the search and switches back to it when nothing is searched.

### Introduces new parent-child relationship logic

-   Prevents that inactive projects can be selected as parents
-   A project cannot be set to inactive if it has an active child
-   A parent project cannot select one of his children as a parent

### Introduces new logic to `ProjectCreateProjectModal`
Adds a new select input in the dialog for creating a project, which includes every active parent and creates a persistent relationship between them on selection.
Licenses and parents are now only retrieved when the dialogue is opened to reduce the amount of data which needs to be loaded for the project list.

### Introducing new logic to `ProjectDetailsModal`
Adds a new select input in the dialog for showing a projects details, which include every possible parent for the current project and creates a persistent relationship between them on selection.
Parents are now only retrieved when the dialogue is opened to reduce the amount of data which needs to be loaded for a project view.

![hierarchical view active](https://user-images.githubusercontent.com/113189967/201305140-39e9879e-9e6e-43a4-9fee-ff7aa60ea720.png)

![hierarchical view inactive](https://user-images.githubusercontent.com/113189967/201305154-ee1e1bc3-8f71-4f01-b4c2-7dcd65239881.png)

![flat view](https://user-images.githubusercontent.com/113189967/201305168-ec95f5a8-1ea1-42db-a4cb-fac27b1579d7.png)

![create project](https://user-images.githubusercontent.com/113189967/201305320-fbd48323-c608-47e1-a602-119385a19c8c.png)

![detail view](https://user-images.githubusercontent.com/113189967/201305212-5caf80cd-88c8-42dd-af53-7d3ab35d3406.png)

Signed-off-by: RBickert <rbt@mm-software.com>

rbt-mm committed:
Implement treegrid for parent-child-relationship

Removed the hierarchical view with subtables in favor of the default
table with an added treegrid to display the hierarchy.
Also includes lazy-loading sub-levels.

Signed-off-by: RBickert <rbt@mm-software.com>

rkg-mm commented on src/views/portfolio/projects/Project.vue:
```diff
@@ -244,8 +244,6 @@
     beforeMount() {
       this.uuid = this.$route.params.uuid;
-    },
-    mounted() {
```
why?

rbt-mm replied:
Are you questioning why I need to set the UUID here or why I removed the `mounted` method and moved `this.initialize` into the `beforeMount` method?

rkg-mm replied:
Why removing the mounted() here and moving it

rbt-mm replied:
`mounted` was accidentally removed, `this.initialize` is now in `mounted` again.
`this.uuid = this.$route.params.uuid` needs to be in `beforeMount`, otherwise the projects uuid will be null in it's detail view because of timing issues.

rbt-mm replied:
Correction: `this.initialize` is necessary in `beforeMount` or else the project won't list its parent in the `ProjectDetailsModal`, unless the project page is refreshed.

rkg-mm commented on src/views/portfolio/projects/ProjectCreateProjectModal.vue:
```diff
@@ -186,10 +199,31 @@
         this.tags = [];
+        this.selectedParent = null;
+        this.availableParents = [{ value: null, text: ''}]
+        this.retrieveParents();
```
is it necessary to do this always? From what I understand this will be done also when the dialog is not shown, so whenever the application is used, and the dialog is rarely used. Can we fetch this data when opening the dialog or dropdown instead?

rbt-mm replied:
It is necessary to reset the available parents, because if a new project is created, it won't show up in the list of selectable parents unless the page is refreshed. But only resetting the available parents after the successful creation of a new project would be better practice.

rkg-mm replied:
@rbt-mm Even better would be to only load the list when its needed, means, when the dialog is being opened to create/update a project. Before it is not necessary but will always be loaded anyways, correct?

rbt-mm replied:
Yes, it will always be loaded on the startup of the project list. This procedure was copied from the already existing `retrievelicenses` method, which are also retrieved via an API call on the startup of the project list.
The method can be changed to only occur on opening the dialog, if wanted.

rkg-mm replied:
Yes change both please in that case, makes no sense to load this data always

rkg-mm commented on src/views/portfolio/projects/ProjectDetailsModal.vue:
```diff
@@ -174,6 +188,39 @@
+            }
+          }
+        }
+        return bool;
```
```suggestion
        return project.children && project.children.some((child) => {
          return child.active || this.hasActiveChild(child);
        });

        return false;
```
Try this to simplify, also I think your logic is flawed in some situations

rbt-mm committed:
Implement dummy children for better lazy-loading

Dummy children are now used instead of loading a level ahead to reduce
API calls and optimize the loading times of a projects next level.

Pagination not working correctly because dummy children need to be
removed and removing from a bootstrap-table decrements the total row
count.

Signed-off-by: RBickert <rbt@mm-software.com>

rbt-mm committed:
Remove dummy children for lazy-loading

Removes dummy children for lazy-loading to avoid the need of removing
data from the table, which caused problems with the pagination.

Now renders the expander button for a project, if it has children which
meet the current filter criteria (e.g. inactive, only applications)

Signed-off-by: RBickert <rbt@mm-software.com>

rbt-mm committed:
Retrieve available parents only when needed

Only fetches the available parents for the `ProjectDetailsModal` and the
`ProjectCreateProjectModal` when the modals are opened, instead of at
the startup of the page, to reduce unnecessary loading of data.

Signed-off-by: RBickert <rbt@mm-software.com>

rbt-mm committed:
Fix checking visibility of wrong column

Signed-off-by: RBickert <rbt@mm-software.com>

rbt-mm committed:
Implement flat view and search

Project list is a hierarchical view per default, can now be switched to
flat view.

Automatically switches to flat view while searching. Disables switching
views while searching.

Signed-off-by: RBickert <rbt@mm-software.com>
